### PR TITLE
Add documentation on Shadow Paths

### PR DESCRIPTION
**Pull request by wenovus**
This is a frequent topic of confusion so want to add some documentation to refer people.

### CODE REVIEW

File: docs/shadow_paths.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+This term is used to describe utilities preferring the `shadow-path` tag instead
+of the `path` tag in the generated GoStructs when they both exist on a field,
+and is therefore used to switch the meaning of the GoStruct from a "config view"
+(intended config+derived state) to a "state view" (applied config+derived
+state).
+
+For example, if a gNMI update for `/interfaces/interface[name="foo"]/state/mtu`
+is unmarshalled into the `Interface` GoStruct at the beginning of this
+documentation with `ygot.SetNode`/`ygot.PreferShadowPath`, then the update would
```
Comment (greg-dennis) [inline docs/shadow_paths.md:79]: why is ygot.SetNode relevant here and can it be dropped? Also, I think it's techincally `ytypes.SetNode`?

File: docs/shadow_paths.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+## Preferring Shadow Paths
+
+This term is used to describe utilities preferring the `shadow-path` tag instead
+of the `path` tag in the generated GoStructs when they both exist on a field,
+and is therefore used to switch the meaning of the GoStruct from a "config view"
+(intended config+derived state) to a "state view" (applied config+derived
+state).
+
+For example, if a gNMI update for `/interfaces/interface[name="foo"]/state/mtu`
```
Comment (greg-dennis) [inline docs/shadow_paths.md:77]: maybe give the converse example where a config update is ignored when ygot.PreferShadowPath is false?

**Review by greg-dennis — CHANGES_REQUESTED** [review]

File: docs/shadow_paths.md (hunk 2 above)
Comment (wenovus) [reply docs/shadow_paths.md]: I didn't do this exactly but expanded on this example and added a description on this "ignoring" behaviour in the previous section.

File: docs/shadow_paths.md (hunk 1 above)
Comment (wenovus) [reply docs/shadow_paths.md]: I'm using it as an example utility. Corrected the name.

**Comment by wenovus** [comment]
Opened an issue for CIFuzz at https://github.com/google/oss-fuzz/issues/9367

File: docs/shadow_paths.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+mirror intended config leaves, which are "config true", or `/config` leaves. Per
+[design.md](design.md#openconfig-path-compression), one of these are compressed
+out depending on the value of the `-prefer_operational_state` generation flag.
+
+Shadow paths are relevant only for compressed GoStructs, and indicate the
+compressed-out `/config` or `/state` YANG `leaf` nodes.
+
+## Problems with Path Compression and how Shadow Paths Help
+
+Path compression leads to the GoStruct not being able to represent both intended
```
Comment (robshakir) [inline docs/shadow_paths.md:62]:
```suggestion
Path compression leads to the GoStruct that ygot generates not being able to represent both intended
```

File: docs/shadow_paths.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+config and applied config at the same time. This leads to two problems:
+
+1.  When subscribing to a non-leaf path, some gNMI clients want to silently
+    ignore the compressed-out paths rather than erroring out due to an
+    unrecognized path.
+2.  Some use cases (e.g. [ygnmi](https://github.com/openconfig/ygnmi#queries))
+    use the same compressed GoStruct for representing either the "config view"
+    (intended config+derived state) combination, or the "state view" (applied
+    config+derived state) combination of leaves. We want to enable such a
+    meaning-switch for certain ygot utilities (e.g. marshalling/unmarshalling).
```
Comment (robshakir) [inline docs/shadow_paths.md:72]: By meaning switch here, do you mean allow switching between `config` views and `state` views?

File: docs/shadow_paths.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+ygot address these issues by,
+
+1.  Always ignoring paths that match a `shadow-path` tag when doing
+    unmarshalling. For example, if a gNMI update for
+    `/interfaces/interface[name="foo"]/config/mtu` is unmarshalled into the
+    `Interface` GoStruct at the beginning of this documentation, then the field
+    will not be populated since it is a shadow path.
+2.  Supporting a `PreferShadowPath` option for some utilities (see section
+    below).
```
Comment (robshakir) [inline docs/shadow_paths.md:82]:
```suggestion
    below). `PreferShadowPath` means that the "shadow" path will be used in preference t the "primary" path annotation.
```

File: docs/shadow_paths.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+## Preferring Shadow Paths
+
+This term is used to describe utilities preferring the `shadow-path` tag instead
```
Comment (robshakir) [inline docs/shadow_paths.md:86]: Which term?

**Review by robshakir — CHANGES_REQUESTED** [review]

File: docs/shadow_paths.md (hunk 4 above)
Comment (wenovus) [reply docs/shadow_paths.md:72]: Yes, changed wording verbatim.

File: docs/shadow_paths.md (hunk 6 above)
Comment (wenovus) [reply docs/shadow_paths.md]: `PreferShadowPath` is behavioural option used to describe utilities preferring...

**Review by robshakir — APPROVED** [review]

**Review by greg-dennis — APPROVED** [review]

**Review by DanG100 — APPROVED** [review]

**Pull request merged**